### DISCUSSION TITLE
Migrating data to new minio cluster without altering lastModify timestamp

### DISCUSSION THREAD

**Discussion by sm-xh** [minio/minio#20296] [Q&A]
Hi,

I am facing an issue during data migration between two MinIO clusters, I am seeking help on how to move the data to new cluster without modifying the timestamp of "lastModified" attribute for objects.

**Old cluster:**
- Platform: Kuberenets
- MinIO Mode: Gateway NAS
- Version: 2020-11-19T23:48:16Z

**New cluster:**
- Platform: 4xUbuntu 22.04 VM
- Type: Mutli-Node Multi-Drive
- Version: 2024-08-03T04:33:23Z

**Steps taken:**
I've tried multiple options to do it which are:
- `mc replicate` - unable to do it since the old cluster API doesn't support this command due to old version: `Unable to add sites for replication. This 'admin' API is not supported by server in 'mode-gateway-nas'.`
- `mc mirror`/`mc cp` (both from S3 API-S3 API, Filesystem-S3 API): data was indeed transfered to new cluster, but all of the objects has lastModified timestamp set to the time of the upload
- `rsync`/`scp `-  as in my understanding, this option is not possible due to MinIO Gateway depreciation, objects needs to be uploaded using S3 API
- `rclone sync/copy `- it was the closest to my goal, but still not what I needed. I run this command with `--metadata` flag, it added `X-Amz-Meta-Btime` and `X-Amz-Meta-Mtime` to metadata section for the objects, however the lastModified timestamp on MinIO is set to the upload date.

**Questions**
I understand that the correct way to do it is using `mc replicate`, however it is not possible in my case. What are the other options, maybe I am missing something? Is it e.g. possible to modify the `lastModfied` timestamp for the object, so I can se it to `X-Amz-Meta-Mtime` after data migration? I checked the documentation, but I didn't find a method that could do so.

**Examples**
```
smxh:~/minio$ mc ls --versions new_minio/mybucket/test.csv
[2024-08-20 19:30:15 CEST]    10B STANDARD null v1 PUT test.csv

smxh:~/minio$ mc ls --versions old_minio/mybucket/test.csv
[2024-02-12 12:03:04 CET]    10B STANDARD test.csv 
```

```
smxh:~/minio$ mc stat old_minio/mybucket/test.csv
Name      : test.csv
Date      : 2024-02-12 12:03:04 CET
Size      : 10 B
ETag      : 7063ccd53b4b79fb2ee6d100fe30af8d
Type      : file
Metadata  :
  Content-Type: text/csv

smxh:~/minio$ mc stat new_minio/mybucket/test.csv
Name      : test.csv
Date      : 2024-08-20 19:30:15 CEST
Size      : 10 B
ETag      : 7063ccd53b4b79fb2ee6d100fe30af8d
Type      : file
Metadata  :
  Content-Type    : text/csv
  X-Amz-Meta-Mtime: 1707735784.286
```

As you can see, after copying/mirroring/syncing the data, timestamp is set to the date of the upload.
Screenshot from MinIO Console:
![image](https://github.com/user-attachments/assets/ff1edad6-1d74-4e48-a680-2997dc5e142c)

**Comment by klauspost** [comment]
S3 API does not provide any way to upload data with a custom timestamp.

**Reply by sm-xh** [reply]
Thanks for the answer. 
So only way to preserve timestamp of last modification of objects is to use `mc replicate`? Is it possible to alter it via e.g. SDK? Is an other option to upgrade old cluster to the version that can access `mc replicate`? I did quite the research, but couldn't find the answer.

**Reply by klauspost** [reply]
No. This version cannot be used for replication. NAS does not support versioning and is way too old.
